parse.go

```diff
@@ -0,0 +1,190 @@
+package ff
+
+import (
+	"bufio"
+	"flag"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// Parse the flags in the flag set from the provided (presumably commandline)
+// args. Additional options may be provided to parse from a config file and/or
+// environment variables in that priority order.
+func Parse(fs *flag.FlagSet, args []string, options ...Option) error {
+	var c Context
+	for _, option := range options {
+		option(&c)
+	}
+
+	if err := fs.Parse(args); err != nil {
+		return errors.Wrap(err, "error parsing commandline args")
+	}
+
+	provided := map[string]bool{}
+	fs.Visit(func(f *flag.Flag) {
+		provided[f.Name] = true
+	})
+
+	if c.configFile == "" && c.configFileFlagName != "" {
+		fs.VisitAll(func(f *flag.Flag) {
+			if f.Name == c.configFileFlagName {
+				c.configFile = f.Value.String()
+			}
+		})
+	}
+
+	if c.configFile != "" && c.configParser != nil {
+		f, err := os.Open(c.configFile)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		c.configParser(f, func(name, value string) error {
+			if fs.Lookup(name) == nil {
+				return errors.Errorf("config file flag %q not defined in flag set", name)
+			}
+
+			if provided[name] {
+				return nil // commandline args take precedence
+			}
+
+			if err := fs.Set(name, value); err != nil {
+				return errors.Wrapf(err, "error setting flag %q from config file", name)
+			}
+
+			return nil
+		})
+	}
+
+	fs.Visit(func(f *flag.Flag) {
+		provided[f.Name] = true
+	})
+
+	if c.envVarPrefix != "" {
+		var errs []string
+		fs.VisitAll(func(f *flag.Flag) {
+			if provided[f.Name] {
+				return // commandline args and config file take precedence
+			}
+
+			var key string
+			{
+				key = strings.ToUpper(f.Name)
+				key = envVarReplacer.Replace(key)
+				key = strings.ToUpper(c.envVarPrefix) + "_" + key
+			}
+			if value := os.Getenv(key); value != "" {
+				for _, individual := range strings.Split(value, ",") {
```

Comment:
Just bumped into this; why do env vars get split by comma?

Reply:
I see. Yes I did have an env var which was a comma separated list of things :) I was under the assumption that string flags and their respective env vars behave similarly. Following up on the PR: https://github.com/peterbourgon/ff/pull/39